### PR TITLE
35287 - Add methods DELETE, PATCH and PUT to Http Client interface and classes

### LINKED ISSUE #35287: Add support for HTTP PATCH requests in /Magento/Framework/HTTP/Adapter/Curl.php
### Summary (*)
Magento has a Curl Adapter which, to my surprise, does not support creating PATCH requests. This results in the class not really being usable if you need to, for example, integrate with a REST API of a third party.

One could argue that in this case the Guzzle Client will suffice, but the problem is that since Magento 2.4.4, Guzzle 7 has been introduced, which introduces new features that are not really backwards compatible with Guzzle 6. For example, implementing the Psr17 and Psr18 specification with guzzle7-adapter vs guzzle6-adapter.
If you want to create modules that are backwards compatible with Magento 2.3 using Guzzle and Psr17/Psr18 implementation, this is not really a viable option anymore.

References:
https://www.php-fig.org/psr/psr-17/
https://www.php-fig.org/psr/psr-18/
https://github.com/php-http/guzzle6-adapter
https://github.com/php-http/guzzle7-adapter

### Examples (*)
-

### Proposed solution
https://github.com/magento/magento2/blob/2.4-develop/lib/internal/Magento/Framework/HTTP/Adapter/Curl.php#L182

From line 182: add something like the following:
```
elseif ($method === \Zend_Http_Client::PATCH) {
            curl_setopt($this->_getResource(), CURLOPT_CUSTOMREQUEST, \Zend_Http_Client::PATCH);
            curl_setopt($this->_getResource(), CURLOPT_POSTFIELDS, $body);
        }
```

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users with no workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [*] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by Nuranto**
### Fixed Issues (if relevant)

1. Fixes magento/magento2#35287

### Questions or comments

Since I added methods to an interface, I guess this is BC.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by Nuranto** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/HTTP/ClientInterface.php [hunk 1]
```diff
@@ -105,6 +113,24 @@ public function get($uri);
...
     public function post($uri, $params);
 
+    /**
+     * Make PATCH request
+     *
+     * @param string $uri full uri
+     * @param array|string $params PATCH fields array or string in case of JSON or XML data
+     * @return void
+     */
+    public function patch($uri, $params);
```
Comment (kandy) [inline lib/internal/Magento/Framework/HTTP/ClientInterface.php]: This is MAJOR breking changes to SPI
Better to move it to separate interfases

btw, include PSR/HttpClient will be best option

File: lib/internal/Magento/Framework/HTTP/Client/Curl.php [hunk 2]
```diff
@@ -229,6 +229,17 @@ public function get($uri)
         $this->makeRequest("GET", $uri);
     }
 
+    /**
+     * Make DELETE request
+     *
+     * @param string $uri uri relative to host, ex. "/index.php"
+     * @return void
+     */
+    public function delete($uri)
```
Comment (engcom-Hotel) [inline lib/internal/Magento/Framework/HTTP/Client/Curl.php]: Please add return type as void
```suggestion
    public function delete($uri): void
```

Please make this appropriate change in other methods as well.

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @Nuranto,

Thanks for the collaboration!

Please merge this branch with the base branch as we do not have access to merge on this fork. 

Also please have a look into the review points below.

Thanks

**Comment by engcom-Hotel** [comment]
We are picking this PR for further development

File: lib/internal/Magento/Framework/HTTP/Client/Curl.php (hunk 2 above)
Comment (nuzil) [reply lib/internal/Magento/Framework/HTTP/Client/Curl.php]: Pls also add parameters types definitions in all methods

File: lib/internal/Magento/Framework/HTTP/Client/Curl.php [hunk 3]
```diff
@@ -246,6 +257,38 @@ public function post($uri, $params)
...
+    /**
+     * Make PUT request
+     *
+     * String type was added to parameter $param in order to support sending JSON or XML requests.
+     *
+     * @param string $uri
+     * @param array|string $params
+     * @return void
+     *
+     * @see \Magento\Framework\HTTP\Client#patch($uri, $params)
```
Comment (nuzil) [inline lib/internal/Magento/Framework/HTTP/Client/Curl.php]: That looks strange for me, I would remove it. This method does not exisits in Magento\Framework\HTTP\Client. Even in ClientInterface
Pls check other locations

**Review by nuzil — CHANGES_REQUESTED** [review]

File: lib/internal/Magento/Framework/HTTP/ClientInterface.php (hunk 1 above)
Comment (Nuranto) [reply lib/internal/Magento/Framework/HTTP/ClientInterface.php]: Not sure what you want me to do here.

Just remove changes from Interface ? Separate interface = ?

Major releases exist for pushing major BC, can't we just wait for it ? (And those who need this quickly can just use a patch, as I did)

Comment (kandy) [reply lib/internal/Magento/Framework/HTTP/ClientInterface.php]: I will recommend to migrate to Psr\Http\Client Interface in you code,with high probability Magento\Framework\HTTP\Client Interface will be deprecated in favor of it
And so, close this PR ;)

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Functional Tests CE, Integration Tests, Semantic Version Checker, Static Tests, Unit Tests, WebAPI Tests

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @Nuranto,

Please look into the failed automated tests and have a look at the below comment:

https://github.com/magento/magento2/pull/38285#discussion_r1508036926

**Comment by engcom-Hotel** [comment]
Hello @Nuranto,

Have you got a chance to check the [review comments here](https://github.com/magento/magento2/pull/38285#pullrequestreview-1910850872)?

Thanks

**Comment by engcom-Hotel** [comment]
Based on this [comment](https://github.com/magento/magento2/pull/38285#discussion_r1508036926), it appears that the implementation is incomplete, and there are failing automated tests. 

Therefore, moving this PR on draft for now.

**Comment by engcom-Hotel** [comment]
Hello @Nuranto,

We appreciate your contribution!

Since we haven't received a response from you for an extended period, we are closing this PR for now.

Whenever you're ready to resume work on this, please feel free to reopen it or request us to do so. We'll be more than happy to assist you.

Thanks once again!

**Pull request closed without merging**